xclim/indices/_agro.py

```diff
@@ -923,123 +977,98 @@ def standardized_precipitation_index(
+    >>> cal_range = ("1990-05-01", "1990-08-31")
     >>> spi_3 = standardized_precipitation_index(
-    ...     pr, pr_cal, freq="MS", window=3, dist="gamma", method="ML"
+    ...     pr,
+    ...     cal_range=cal_range,
+    ...     freq="MS",
+    ...     window=3,
+    ...     dist="gamma",
+    ...     method="ML",
     ... )  # Computing SPI-3 months using a gamma distribution for the fit
+    >>> # Fitting parameters can also be obtained ...
+    >>> params = _compute_spx_fit_params(
+    ...     pr,
+    ...     cal_range,
+    ...     freq="MS",
+    ...     window=3,
+    ...     dist="gamma",
+    ...     method="ML",
+    ... )  # First getting params
+    >>> # ... and used as input
+    >>> spi_3 = standardized_precipitation_index(pr, params=params)
 
     References
     ----------
     :cite:cts:`mckee_relationship_1993`
-
     """
-    # "WPM" method doesn't seem to work for gamma or pearson3
-    dist_and_methods = {"gamma": ["ML", "APP"], "fisk": ["ML", "APP"]}
-    if dist not in dist_and_methods:
-        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
-    if method not in dist_and_methods[dist]:
-        raise NotImplementedError(
-            f"The method `{method}` is not supported for distribution `{dist}`."
+    uses_input_params = params is not None
+    if cal_range and uses_input_params:
+        raise ValueError(
+            "Inputing both calibration dates (`cal_range`) and calibration parameters (`params`) is not accepted,"
+            "input only one or neither of those options (the latter case reverts to default behaviour which performs a calibration"
+            "with the full input dataset)."
         )
 
-    # calibration period
-    cal_period = pr_cal.time[[0, -1]].dt.strftime("%Y-%m-%dT%H:%M:%S").values.tolist()
-
-    # Determine group type
-    if freq == "D" or freq is None:
-        freq = "D"
-        group = "time.dayofyear"
-    else:
-        _, base, _, _ = parse_offset(freq)
-        if base in ["M"]:
-            group = "time.month"
-        else:
-            raise NotImplementedError(f"Resampling frequency `{freq}` not supported.")
-
-    # Resampling precipitations
-    if freq != "D":
-        pr = pr.resample(time=freq).mean(keep_attrs=True)
-        pr_cal = pr_cal.resample(time=freq).mean(keep_attrs=True)
-
-        def needs_rechunking(da):
-            if uses_dask(da) and len(da.chunks[da.get_axis_num("time")]) > 1:
-                warnings.warn(
-                    "The input data is chunked on time dimension and must be fully rechunked to"
-                    " run `fit` on groups ."
-                    " Beware, this operation can significantly increase the number of tasks dask"
-                    " has to handle.",
-                    stacklevel=2,
-                )
-                return True
-            return False
-
-        if needs_rechunking(pr):
-            pr = pr.chunk({"time": -1})
```

Comment:
Before, I would rechunk everything, then compute parameters and compute spi values.

Now, I compute parameters, rechunk only the concerned subset, and then move on and compute spi values with the original chunked `pr`.

This removes warnings about out-of-index slices or something like this.  However, it greatly increases the number of dask tasks (from about ~50% of original SPI tasks to 150% of original SPI tasks)